File: frontend/src/types/report.ts
import type { CashierShift, SaleOrder } from "@/types/pos";

export interface DailySalesSummary {
  date: string;
  date_thai: string;
  total_orders: number;
  total_amount: number;
  total_vat: number;
  total_discount: number;
  avg_order_value: number;
  by_payment_method: Record<string, number>;
}

export interface SalesRangeSummary {
  date_from: string;
  date_to: string;
  days: DailySalesSummary[];
  grand_total_orders: number;
  grand_total_amount: number;
  grand_total_vat: number;
  grand_total_discount: number;
  by_payment_method: Record<string, number>;
}

export interface TopProductItem {
  product_id: string;
  product_name: string;
  sku: string;
  total_qty: number;
  total_amount: number;
  order_count: number;
}

export interface TopProductsReport {
  date_from: string;
  date_to: string;
  items: TopProductItem[];
}

export interface HourlySales {
  hour: number;
  total_orders: number;
  total_amount: number;
}

export interface DashboardStats {
  today_orders: number;
  today_sales: number;
  today_vat: number;
  today_avg_order: number;
  open_shifts_count: number;
  low_stock_count: number;
  total_products: number;
  compared_yesterday_pct: number | null;
}

export interface ShiftSummary {
  shift: CashierShift;
  sales: SaleOrder[];
  voided: SaleOrder[];
  cashier_name: string;
  branch_name: string;
  location_name: string;
  by_payment_method: Record<string, number>;
  daily_summary: DailySalesSummary;
}

File: backend/app/schemas/report.py
from __future__ import annotations

from datetime import date
from decimal import Decimal

from pydantic import ConfigDict

from app.schemas import BaseSchema
from app.schemas.pos import SaleOrderRead, ShiftRead


class DailySalesSummary(BaseSchema):
    date: date
    date_thai: str
    total_orders: int
    total_amount: Decimal
    total_vat: Decimal
    total_discount: Decimal
    avg_order_value: Decimal
    by_payment_method: dict[str, Decimal]


class SalesRangeSummary(BaseSchema):
    date_from: date
    date_to: date
    days: list[DailySalesSummary]
    grand_total_orders: int
    grand_total_amount: Decimal
    grand_total_vat: Decimal
    grand_total_discount: Decimal
    by_payment_method: dict[str, Decimal]


class TopProductItem(BaseSchema):
    product_id: str
    product_name: str
    sku: str
    total_qty: Decimal
    total_amount: Decimal
    order_count: int


class TopProductsReport(BaseSchema):
    date_from: date
    date_to: date
    items: list[TopProductItem]


class HourlySales(BaseSchema):
    hour: int
    total_orders: int
    total_amount: Decimal


class ShiftSummary(BaseSchema):
    shift: ShiftRead
    sales: list[SaleOrderRead]
    voided: list[SaleOrderRead]
    daily_summary: DailySalesSummary
    by_payment_method: dict[str, Decimal]
    cashier_name: str
    branch_name: str
    location_name: str


class DashboardStats(BaseSchema):
    today_orders: int
    today_sales: Decimal
    today_vat: Decimal
    today_avg_order: Decimal
    open_shifts_count: int
    low_stock_count: int
    total_products: int
    compared_yesterday_pct: Decimal | None = None

    model_config = ConfigDict(from_attributes=True)
